2996_reflection/reflection.md

```diff
@@ -1886,9 +1888,19 @@ namespace std::meta {
 
 This metafunction produces a reflection of the value returned by a call expression.
 
-Letting `F` be the entity reflected by `target`, and `A_0, ..., A_n` be the sequence of entities reflected by the values held by `args`: if the expression `F(A_0, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, and if every value in `args` is a reflection of a constant value, then `reflect_invoke(target, args)` evaluates to a reflection of the constant value `F(A_0, ..., A_N)`.
+If any value held by `args` is not a reflection of a constant value, then `reflect_invoke(target, args)` is not a constant expression. Otherwise it is evaluated by the following rules, where `F` is the entity reflected by `target` and `A_0, ..., A_n` is the sequence of values reflected by the elements indexed by `args`.
 
-For all other invocations, `reflect_invoke(target, args)` is ill-formed.
+If `F` is a non-static member function or non-static member function template, `A_0` is a reflection of a value not of a pointer type, and the expression `A_0.F(A_1, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, then `reflect_invoke(target, args)` evaluates to a reflection of the value `A_0.F(A_1, ..., A_N)`.
+
+If `F` is a non-static member function or non-static member function template, `A_0` is a reflection of a value of a pointer type, and the expression `A_0->F(A_1, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, then `reflect_invoke(target, args)` evaluates to a reflection of the value `A_0->F(A_1, ..., A_N)`.
+
+If `F` is a pointer to a non-static member function, `A_0` is a reflection of a value not of a pointer type, and the expression `A_0.*F(A_1, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, then `reflect_invoke(target, args)` evaluates to a reflection of the value `A_0.*F(A_1, ..., A_N)`.
+
+If `F` is a pointer to a non-static member function, `A_0` is a reflection of a value of a pointer type, and the expression `A_0->*F(A_1, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, then `reflect_invoke(target, args)` evaluates to a reflection of the value `A_0->*F(A_1, ..., A_N)`.
+
+If the expression `F(A_0, ..., A_N)` is a well-formed constant expression evaluating to a type that is not `void`, then `reflect_invoke(target, args)` evaluates to a reflection of the constant value `F(A_0, ..., A_N)`.
+
+For all other invocations, `reflect_invoke(target, args)` is not a constant expression.
```

Comment:
We should just say that `reflect_invoke(^F, ^{Args...})` evaluates as `std::invoke(F, Args...)`.

Reply:
I think we still need special cases for `^S::fn`, since you otherwise would need to spell it `&S::fn`, but that might otherwise work.

Reply:
Yes, it would evaluate to a reflection of a constant value equal to the result of `std::invoke(F, Args...)`.

Reply:
Actually, we also need a special case for where `F` is a function template (which I assume we want to allow).

Reply:
Oh actually, using `std::invoke` would also rule out constructors. I'd rather keep the language as is for now, and look for opportunities to shorten it later.

Reply:
So the difference between `reflect_invoke(^F, {^Args...})` and `std::invoke(F, Args...)`:

1. you can `reflect_invoke` on a constructor, which you cannot `std::invoke` - there's no actual way to name the constructor. 
2. you can `invoke(&C::mem, std::ref(c))` but you cannot `reflect_invoke(^C::mem, ^std::ref(c))`. 
3. you cannot `reflect_invoke` a function that returns `void`

And possibly, but I'm not sure this is intended:

4. you can `invoke(&C::mem, std::make_unique<C>())` but you cannot `reflect_invoke` on that.

I'd rather describe `reflect_invoke` in terms of a delta from `std::invoke` since people are familiar with that one and this is just prose, cause the important parts are the differences between the two - specifically which ones are on purpose. 

Also why not allow returning `void`?